### PR TITLE
Unifying Pre- and Postprocessing

### PR DESCRIPTION
**Pull request by FynnBe**

### CODE REVIEW

File: supported_formats_and_operations.md [hunk 1]
```diff
@@ -23,32 +23,40 @@ Which consumer software supports which format?
...
   - `kwargs`
     - `gain` multiplicative factor
     - `offset` additive factor
     - `axes` the subset of axes to scale jointly. For example `xy` to scale the two image axes for 2d data jointly. The batch axis (`b`) is not valid here.
   - `reference_implementation`
-- `scale_min_max` normalize the tensor to range 0, 1
+- `scale_mean_variance` scale the tensor s.t. its mean and variance match a reference tensor 
+  - `kwargs`
+    - `mode` one of `per_dataset` or `per_sample` (for fixed mean and variance use `scale_linear`)
+    - `reference_tensor` name of tensor to match
```
Comment (constantinpape) [inline supported_formats_and_operations.md:48]: This doesn't make much sense in pre-processing.

File: supported_formats_and_operations.md [hunk 2]
```diff
@@ -23,32 +23,40 @@ Which consumer software supports which format?
...
+  - `reference_implementation`
+- `scale_range` normalize the tensor with percentile normalization
   - `kwargs`
-    - `mode` can be one of `per_sample` (min and max are computed for each sample individually), `per_dataset` (min and max are computed for the entire dataset)
+    - `mode` can be one of `per_sample` (percentiles are computed for each sample individually), `per_dataset` (percentiles are computed for the entire dataset). For a fixed scaling use `scale linear`.
     - `axes` the subset of axes to normalize jointly. For example `xy` to normalize the two image axes for 2d data jointly. The batch axis (`b`) is not valid here.
+    - `min_percentile` the lower percentile used for normalization, in range 0 to 100. Default value: 0.
+    - `max_percentile` the upper percentile used for normalization, in range 0 to 100. Has to be bigger than `upper_percentile`. Default value: 100.
+    - `reference_tensor` tensor name to compute the percentiles from. Default: The tensor itself. If `mode`==`per_dataset` this needs to be the name of an input tensor.
```
Comment (constantinpape) [inline supported_formats_and_operations.md:56]: This doesn't make much sense in pre-processing either.

**Review by constantinpape — COMMENTED** [review]
The problem is that the `reference_tensor` in `scale_range` and `scale_mean_variance` only makes sense in post-processing. So it is a bit awkward to unify pre-and-postprocessing and then have special cases for it in each function.

**Comment by constantinpape** [comment]
@FynnBe let me know when you have made all the changes and I should have a look again.

**Comment by FynnBe** [comment]
> @FynnBe let me know when you have made all the changes and I should have a look again.

I think it's all up to date. I left `binarize` and `sigmoid` in the common processing's, as they might not be as useful for the input, but wouldn't have any other kwargs...

File: supported_formats_and_operations.md [hunk 3]
```diff
@@ -23,95 +23,96 @@ Which consumer software supports which format?
...
+  - `kwargs` None
   - `reference_implementation`
 - `zero_mean_unit_variance` normalize the tensor to have zero mean and unit variance
   - `kwargs`
     - `mode` can be one of `fixed` (fixed values for mean and variance), `per_sample` (mean and variance are computed for each sample individually), `per_dataset` (mean and variance are computed for the entire dataset)
     - `axes` the subset of axes to normalize jointly. For example `xy` to normalize the two image axes for 2d data jointly. The batch axis (`b`) is not valid here.
     - `mean` the mean value(s) to use for `mode == fixed`. For example `[1.1, 2.2, 3.3]` in the case of a 3 channel image where the channels are not normalized jointly.
     - `std` the standard deviation values to use for `mode == fixed`. Analogous to `mean`.
-    - `[eps]` epsilon for numeric stability: `out = (tensor - mean) / (std + eps)`.
+    - `[eps]` epsilon for numeric stability: `out = (tensor - mean) / (std + eps)`. Default value: 10e-7.
```
Comment (constantinpape) [inline supported_formats_and_operations.md:54]: Nitpicky: It's either `10^-7` or `1e-7`.

File: supported_formats_and_operations.md [hunk 4]
```diff
@@ -23,95 +23,96 @@ Which consumer software supports which format?
...
-|  `min_max`                 | ?       | ?          | ?    |
-|  `percentile`              | ?       | ?          | ?    |
-| `scale_linear`             | ?       | ?          | ?    |
+|  `scale_linear`            | ?       | ?          | ?    |
+|  `sigmoid`                 | yes     | ?          | ?    |
 |  `zero_mean_unit_variance` | yes     | ?          | ?    |
 
 
-# Postprocessing
+## Preprocessing
```
Comment (constantinpape) [inline supported_formats_and_operations.md:70]: What about `scale_mean_variance`? I think you forgot it.

File: supported_formats_and_operations.md [hunk 5]
```diff
@@ -23,95 +23,96 @@ Which consumer software supports which format?
...
+| preprocesing               | ilastik | deepImageJ | Fiji |
+| -------------------------- | ------- | ---------- | ---- |
+|  `scale_range`             | ?       | ?          | ?    |
+
+
+## Postprocessing
+
+Additional postprocessing operations.
+
+- `scale_range` normalize the tensor with percentile normalization
```
Comment (constantinpape) [inline supported_formats_and_operations.md:96]: We should add something like "Same keyword arguments as preprocessing:scale_range + reference_tensor"

File: supported_formats_and_operations.md [hunk 6]
```diff
@@ -23,95 +23,96 @@ Which consumer software supports which format?
...
   - `kwargs`
-    - `mode` one of `per_dataset` or `per_sample` (for fixed mean and variance use `scale_linear`)
-    - `reference_input` name of the input tensor to use as reference
+    - same as preprocessing
+    - `reference_tensor` tensor name to compute the percentiles from. Default: The tensor itself. If `mode`==`per_dataset` this needs to be the name of an input tensor.
   - `reference_implementation`
-- `scale_min_max` scale the tensor s.t. its min and max match a reference tensor
+- `scale_mean_variance` scale the tensor s.t. its mean and variance match a reference tensor
```
Comment (constantinpape) [inline supported_formats_and_operations.md:101]: Also refere to the preprocessing here.

File: supported_formats_and_operations.md [hunk 7]
```diff
@@ -23,95 +23,96 @@ Which consumer software supports which format?
...
-| preprocesing          | ilastik | deepImageJ | Fiji |
+| postprocesing          | ilastik | deepImageJ | Fiji |
 | --------------------- | ------- | ---------- | ---- |
-|  `clip`               | no      | ?          | yes  |
-|  `binarize`           | no      | ?          | ?    |
-|  `sigmoid`            | yes     | ?          | ?    | 
-|  `scale_mean_variance`| no      | ?          | yes  |
-|  `scale_min_max`      | no      | ?          | yes  |
-|  `scale_linear`       | no      | ?          | ?    |
+|  `scale_mean`              | ?       | ?          | ?    |
```
Comment (constantinpape) [inline supported_formats_and_operations.md:114]: Should be `scale_man_variance`

**Review by constantinpape — CHANGES_REQUESTED** [review]
I think you forgot `scale_mean_variance` in preprocessing.

File: supported_formats_and_operations.md (hunk 5 above)
Comment (FynnBe) [reply supported_formats_and_operations.md:96]: there is ... 

```
`kwargs`: 
  - same as preprocessing
  - `reference_tensor`
``` 
or am I misunderstanding you here?

Comment (constantinpape) [reply supported_formats_and_operations.md:96]: ah sorry, I just overlooked that.

**Comment by FynnBe** [comment]
> I think you forgot scale_mean_variance in preprocessing.

`zero_mean_unit_variance` is in the Pre- and Postprocessing section...

**Comment by constantinpape** [comment]
> `zero_mean_unit_variance` is in the Pre- and Postprocessing section...

But then `same as preprocessing` in the kwargs is ambiguous and we need to refer to `zero_mean_unit_variance` there.

File: supported_formats_and_operations.md (hunk 6 above)
Comment (FynnBe) [reply supported_formats_and_operations.md:101]: should `scale_mean_variance` exist for preprocessing? It is different from the `zero_mean_unit_variance`.

**Comment by FynnBe** [comment]
> But then same as preprocessing in the kwargs is ambiguous and we need to refer to zero_mean_unit_variance there.
same as preprocessing is now only specified for `scale_range`

**Comment by FynnBe** [comment]
actually for `scale_mean_variance` a `mode=fixed`  would also makes sense, but one can achieve the same with `zero_mean_unit_variance` + `scale_linear`. Therefore I would leave it at that for now.

**Review by constantinpape — COMMENTED** [review]
Ok, I think I understand it know and it should all be correct:
For `scale_mean_variance` we only need to know the `reference_tensor` as kwarg, so we actually don't need to refer to any other kwargs.

**Review by constantinpape — APPROVED** [review]

**Comment by FynnBe** [comment]
> Ok, I think I understand it know and it should all be correct:
> For `scale_mean_variance` we only need to know the `reference_tensor` as kwarg, so we actually don't need to refer to any other kwargs.

The `mode` (`per_sample` or `per_dataset`) is required as well to determine how to determine mean and reference from the reference.

**Comment by constantinpape** [comment]
> The `mode` (`per_sample` or `per_dataset`) is required as well to determine how to determine mean and variance from the reference.

True, but we have that already.
I just wanted to point out that we don't need to refer to any other kwargs.

This is good to go now and I will merge.

**Pull request merged**